netperf_tool/src/server_node.hpp

```diff
@@ -0,0 +1,151 @@
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "netperf_tool_interfaces/msg/bytes.hpp"
+#include "netperf_tool_interfaces/srv/get_results.hpp"
+
+#include "rclcpp/executors.hpp"
+#include "rclcpp/node.hpp"
+#include "rclcpp/serialization.hpp"
+#include "rclcpp/serialized_message.hpp"
+#include "rclcpp/subscription.hpp"
+
+namespace netperf_tool
+{
+struct ServerMessageInfo
+{
+  /// Timestamp of when the message was received.
+  rclcpp::Time reception_time;
+  /// Message ids.
+  size_t id;
+  /// Single trip latency of the message.
+  std::chrono::nanoseconds latency;
+  /// Serialized message size.
+  size_t serialized_size;
+};
+
+/// Raw data with info about received messages.
+struct ServerCollectedInfo
+{
+  /// Collected info of each message.
+  std::vector<ServerMessageInfo> message_infos;
+};
+
+/// The results of an experiment.
+struct ServerResults
+{
+  /// Statistics, calculated when client makes a request.
+  netperf_tool_interfaces::srv::GetResults::Response statistics;
+  /// Raw message data collected from the received messages.
+  ServerCollectedInfo collected_info;
+};
+
+/// Node used as server side of a ros2 netperf experiment.
+class ServerNode : public rclcpp::Node
+{
+public:
+  /// Construct.
+  /**
+   * \param[in] options Node options used to construct the node.
+   * \param[in] sub_qos Qos profile to create the publisher used in the experiment.
+   */
+  explicit ServerNode(
+    const rclcpp::NodeOptions & options,
+    const rclcpp::QoS & sub_qos);
+
+  /// Callback executed when a message is received.
+  /**
+   * \param[in] serialized_msg Serialized message received from client.
+   * \param[in] msg_info Message info provided by the middleware.
+   */
+  void handle_msg(
+    const rclcpp::SerializedMessage & serialized_msg,
+    const rclcpp::MessageInfo & msg_info);
+
+  /// Callback executed when client requests experiment results.
+  /**
+   * \param[in] req Request made by client.
+   * \param[in] rep Response sent to client.
+   */
+  void
+  handle_get_results_request(
+    netperf_tool_interfaces::srv::GetResults::Request::SharedPtr req,
+    netperf_tool_interfaces::srv::GetResults::Response::SharedPtr rep);
+
+  /// Map from publisher gid to results.
+  using PubGidToResultsMap = std::unordered_map<std::string, ServerResults>;
+
+  /// Extract available results of experiments.
+  /**
+   * This method will clear the currently stored results, so calling it twice in a row
+   * will result in empty results the second time.
+   */
+  PubGidToResultsMap
+  extract_results();
+
+  /// Extract publishers gids of new clients found.
+  /**
+   * This method will clear the currently stored gid, so calling it twice in a row
+   * will result in empty results the second time.
+   */
+  std::vector<std::string>
+  extract_new_clients();
+
+  /// Extract until results of an experiment are available or a new client is found.
```

Comment:
nit: docstring?

Reply:
fixed in https://github.com/ros2/netperf/pull/1/commits/257523aa13b6b81126a5a804648b054b19e39939.